### ISSUE TITLE
Broadcasting and Additional Dimensions for pairwise_distance

### ISSUE THREAD

**Issue by elbaro** [pytorch/pytorch#15918]
## 🚀 Feature

From docs, `F.pairwise_distance` supports `[N,D]x[N,D]=[N]`.
Please add the support for `[(N1,N2,..),D]x[(N1,N2,..),D]=[N1,N2,..]` with broadcasting.

## Motivation

I am trying to implement a triplet loss.
From the network I have a embedding `[Batch, D]` and mean vectors `[num_cls, D]`.
I need distances between pairs, `[Batch,num_cls]`. However `pdist` and `pairwise_distance` do not cover this.

If `pairwise_distance` supports multiple dimension, it is possible to unsqueeze them to `[Batch, 1, D]x[1,num_cls,D]` and get `[Batch,num_cls]`.

**Comment by xmnlab** [comment]
hi @elbaro just to understand better this issue.

As I could understand:

1. changes should be on functional.pairwise_distance [1]
2. x1 and x2 should accept Tensor or Tuple
3. x1 and x2 types should be equal.
4. x1 and x2 if tuple should have the same number of elements 
5. for x1 and x2 tuples, results should be a list of the result for each (x1_i and x2_i)

[1] https://github.com/pytorch/pytorch/blob/da753b7ccf22ad75c6b923b5f059b703e7081426/torch/nn/functional.py#L2720

**Comment by fmassa** [comment]
I'll put this on a hold for now, as it's not clear to me if it's already implemented in https://github.com/pytorch/pytorch/pull/12302 or not.

**Comment by fmassa** [comment]
My understanding when I first read the issue was that all that was needed to be doing was to reshape the tensors to be `(N1,N2,...) x D`, perform `pdist` and then reshape back the results, but maybe I misunderstood something, and also #12302 already implements batched `pdist`, so it might address this issue

**Comment by xmnlab** [comment]
OK I have subscribed to this issue to receive any update.

**Comment by elbaro** [comment]
I edited the issue. Basically it's the same as that `nn.Linear` supports `(N,*,D)`. From docs `pdist` with input `[N,D]` returns the `N*(N-1)/2` distances, and `pairwise_distance` with input `[N,D], [N,D]` returns `N` distances. I have two inputs `[N,D], [M,D]` and want to calculate `N*M` distances.

One way to implement it is generalizing `pairwise_distance` to support any dimension and feed `[N,1,D], [1, M, D]` and they are broadcasted to `[N,M,D], [N,M,D]` resulting in `[N,M]`. The only change is the supported tensor shape.

However I have no preference on specific interfaces.

**Comment by rgemulla** [comment]
> From the network I have a embedding `[N, D]` and mean vectors `[M, D]`.
> I need `N*M` distances between them. However `pdist` and `pairwise_distance` do not cover this.

Have a look at `cdist`; from my understanding, it's what you want to do.

**Comment by elbaro** [comment]
Great, the newly added `cdist` does this.
Feel free to close (or leave if anyone still wants a generalized pairwise_distance).
